transports/quic/src/muxer.rs

```diff
@@ -254,12 +276,9 @@ impl AsyncRead for Substream {
         let mut bytes = 0;
         let mut pending = false;
         loop {
-            if buf.is_empty() {
-                break;
-            }
             let chunk = match chunks.next(buf.len()) {
```

Comment:
> probably going to often hit this case.

Really? Most protocols will wrap the substream in some kind of `Framed` to parse chunks of bytes as messages, which means whatever buffer they have needs to be at least large enough to hold one message, parse it and clear the buffer again.

I've not measured it but my intuition would have been that the common case will be that we are out of chunks and not that the buffer is full.

I don't really mind much which way we are coding this, if you feel strongly about the other style, I am happy with it too. IMO, we ideally upstream a change to `quinn` to return `None` instead of a zero-sized chunk.

Reply:
Actually, after reading the code again, I think `Ok(None)` here means that the stream is finished: https://github.com/quinn-rs/quinn/blob/3a2558258034e60989bbc199d4d8b0b7297ee269/quinn-proto/src/connection/streams/recv.rs#L239-L280

Should we handle this separately? Effectively, we will return `Ok(0)` here with how the control flow is currently set up so it may not matter.

Reply:
> Really? Most protocols will wrap the substream in some kind of `Framed` to parse chunks of bytes as messages, which means whatever buffer they have needs to be at least large enough to hold one message, parse it and clear the buffer again.

That's a Good point! I guess it's not obvious we're often hitting this case. I made a quick hack to see how often the smoke tests hit this (it's used length prefixed data): https://gist.github.com/MarcoPolo/8b04abc59e3b0b03a1bee9490d296fb7.

Here's the result:

```
❯ cargo test --test smoke --features async-std -- --nocapture --test async_std_smoke
...
running 1 test
Found empty buf. Times: 1 out of 1. (loop run: 2)
Found empty buf. Times: 2 out of 2. (loop run: 2)
...
Found empty buf. Times: 28 out of 41. (loop run: 2)
test async_std_smoke ... ok
```

I omitted some lines, but 28 out of the 41 times we called this function ended with us seeing an empty buf. This seems surprising to me for the reason you bring up. And I haven't looked into this further, but my guess is if there is exactly N bytes left and you know you are reading N bytes and you [allocate a buf of N bytes](https://github.com/kpp/rust-libp2p/blob/master/core/src/upgrade/transfer.rs#L123), then you'll fill it up and will get an empty buf on the next loop.

If the protocol takes care of reusing buffers that are bigger than the expected bytes you need to read, you won't hit this. But it seems like in this one experiment we hit it pretty often.

Reply:
Nice investigation!

Happy for it to go back to the old implementation.

Reply:
> Actually, after reading the code again, I think `Ok(None)` here means that the stream is finished: https://github.com/quinn-rs/quinn/blob/3a2558258034e60989bbc199d4d8b0b7297ee269/quinn-proto/src/connection/streams/recv.rs#L239-L280
> 
> Should we handle this separately? Effectively, we will return `Ok(0)` here with how the control flow is currently set up so it may not matter.

I think `Ok(None)` can also be returned after we have already iterated through the loop at least once and filled our buffer, in which case we return `Ok` with the bytes read in those previous iterations.  
Is that what you meant or did I misunderstand your point?

---

Thanks for the investigation @MarcoPolo! Went back to the old implementation but added an extra comment: a38f0e95afc3edc0e9f1ba4003dac52024911289.